homeassistant/components/tplink/entity.py

```diff
@@ -274,55 +255,92 @@ def _entities_for_device_and_its_children[_E: CoordinatedTPLinkEntity](
+            if child_coordinators:
+                child_coordinator = child_coordinators[idx]
+            else:
+                child_coordinator = coordinator
             entities.extend(
                 _entities_for_device(
                     child,
-                    coordinator=coordinator,
+                    coordinator=child_coordinator,
                     feature_type=feature_type,
                     entity_class=entity_class,
                     parent=device,
                 )
             )
 
-    entities.extend(
-        _entities_for_device(
-            device,
-            coordinator=coordinator,
-            feature_type=feature_type,
-            entity_class=entity_class,
-        )
-    )
-
     return entities
 
 
+def _category_for_feature(feature: Feature | None) -> EntityCategory | None:
+    """Return entity category for a feature."""
+    # Main controls have no category
+    if feature is None or feature.category is Feature.Category.Primary:
+        return None
+
+    if (
+        entity_category := FEATURE_CATEGORY_TO_ENTITY_CATEGORY.get(feature.category)
+    ) is None:
+        _LOGGER.error("Unhandled category %s, fallback to DIAGNOSTIC", feature.category)
+        entity_category = EntityCategory.DIAGNOSTIC
+
+    return entity_category
+
+
 def _description_for_feature[_D: EntityDescription](
-    desc_cls: type[_D], feature: Feature, **kwargs: Unpack[EntityDescriptionExtras]
+    desc_cls: type[_D],
+    feature: Feature,
 ) -> _D:
     """Return description object for the given feature.
 
     This is responsible for setting the common parameters & deciding based on feature id
     which additional parameters are passed.
     """
 
-    # Disable all debug features that are not explicitly enabled.
-    if "entity_registry_enabled_default" not in kwargs:
-        kwargs["entity_registry_enabled_default"] = (
-            feature.category is not Feature.Category.Debug
-        )
-
-    return desc_cls(
-        key=feature.id,
-        translation_key=feature.id,
-        name=feature.name,
-        **kwargs,
+    kwargs: dict = {}
+    # Key can be overridden for legacy support
+    if (entity_extras := ENTITY_EXTRAS.get(feature.id)) and (
+        key := entity_extras.get("key")
+    ):
+        kwargs["key"] = key
+    else:
+        kwargs["key"] = feature.id
+    # Use translation keys for all features configured in extras
+    if entity_extras:
+        kwargs["translation_key"] = feature.id
+        kwargs["state_class"] = entity_extras.get("state_class")
+        kwargs["device_class"] = entity_extras.get("device_class")
+    else:
+        kwargs["name"] = feature.name
+        kwargs["icon"] = feature.icon
+
+    kwargs["entity_registry_enabled_default"] = (
+        feature.category is not Feature.Category.Debug
     )
+    kwargs["entity_category"] = _category_for_feature(feature)
+
+    if feature.type == Feature.Type.Sensor:
```

Comment:
```suggestion
    if feature.type is Feature.Type.Sensor:
```

if both sides are an enum